Xamarin.Essentials/TextToSpeech/TextToSpeech.android.cs

```diff
@@ -0,0 +1,246 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Android.OS;
+using Android.Speech.Tts;
+using AndroidTextToSpeech = Android.Speech.Tts.TextToSpeech;
+
+namespace Xamarin.Essentials
+{
+    public static partial class TextToSpeech
+    {
+        const int maxSpeechInputLengthDefault = 4000;
+        internal const float PlatformSpeakRateMin = 0f;
+        internal const float PlatformSpeakRateMax = 2f;
+        internal const float PlatformPitchMin = 0f;
+        internal const float PlatformPitchMax = 2f;
+
+        static WeakReference<TextToSpeechImplementation> textToSpeechRef = null;
+        static SemaphoreSlim semaphore;
+
+        static TextToSpeechImplementation GetTextToSpeech()
+        {
+            TextToSpeechImplementation tts = null;
+            if (textToSpeechRef == null || !textToSpeechRef.TryGetTarget(out tts))
+            {
+                tts = new TextToSpeechImplementation();
+                textToSpeechRef = new WeakReference<TextToSpeechImplementation>(tts);
+            }
+
+            return tts;
+        }
+
+        internal static async Task PlatformSpeakAsync(string text, SpeakSettings settings, CancellationToken cancelToken = default)
+        {
+            if (semaphore == null)
+                semaphore = new SemaphoreSlim(1, 1);
+
+            using (var textToSpeech = GetTextToSpeech())
+            {
+                var max = maxSpeechInputLengthDefault;
+                if (Platform.HasApiLevel(BuildVersionCodes.JellyBeanMr2))
+                    max = AndroidTextToSpeech.MaxSpeechInputLength;
+
+                try
+                {
+                    await semaphore.WaitAsync(cancelToken);
+                    await textToSpeech.SpeakAsync(text, max, settings, cancelToken);
+                }
+                finally
+                {
+                    if (semaphore.CurrentCount == 0)
+                        semaphore.Release();
+                }
+            }
+        }
+
+        internal static Task<IEnumerable<Locale>> PlatformGetLocalesAsync()
+        {
+            using (var textToSpeech = GetTextToSpeech())
+                return textToSpeech.GetLocalesAsync();
+        }
+    }
+
+    internal class TextToSpeechImplementation : Java.Lang.Object, AndroidTextToSpeech.IOnInitListener,
+#pragma warning disable CS0618
+        AndroidTextToSpeech.IOnUtteranceCompletedListener
+#pragma warning restore CS0618
+    {
+        AndroidTextToSpeech tts;
+        TaskCompletionSource<bool> tcs;
+        TaskCompletionSource<object> tcsUtterances;
```

Comment:
linker may get confused with object here....

Reply:
You mean to just use `bool` or something?